### ISSUE TITLE
retained dedup?

### ISSUE THREAD

**Issue by behrad** [mcollina/aedes#126]
retained packets are disabled for dedup [here](https://github.com/mcollina/aedes/blob/master/lib/handlers/subscribe.js#L146) Why?

Currently retained packets are delivered twice if a `clean=false` client does:

CONNECT
SUBSCRIBE

let's make this configurable for the application atleast @mcollina to choose if they want to dedup retained packets or not.

**Comment by mcollina** [comment]
technically those are two subscribes, so I think they should be delivered twice. You should not be calling SUBSCRIBE again, if the `sessionPresent` part of the CONNACK packet is set.

**Comment by behrad** [comment]
> You should not be calling SUBSCRIBE again, if the sessionPresent part of the CONNACK packet is set

Good point, I should recheck the client's behavior on our side 👍 

Can you explain what is the job of `dedup` @mcollina ? What's `brokerCounter`?
it seems `dedup` is only deduping for connected clients in current running broker instance. (it doesn't survive broker restarts)

**Comment by mcollina** [comment]
It's to handle overlapping subscriptions (`a/#` `a/b` ). We need to dedupe, as there are two handlers registered in mqemitter.

**Comment by behrad** [comment]
It could have a finer qualified function name then ;)
I still like to make dedup retained configurable for the application (as a non-standard option) so that client's don't receive same retained packet on each CONNECT. Can I create a PR if you don't mind @mcollina ?

**Comment by mcollina** [comment]
@behrad maybe we should not dedupe, or maybe do something different there. Would you mind assembling a unit test for what you think it should happen?

**Comment by GavinDmello** [comment]
Ideally, on every Subscribe you should get the retained packet for that topic.

**Comment by behrad** [comment]
In my environment there's a high CONNECT/DISCONNECT rate (a few hundred per second)
When we publish a retained message, all the bandwidth is eaten by retained forwards. If I could turn the `dedup_retained_on_connect` switch on, it would help me so much.


> Ideally, on every Subscribe you should get the retained packet for that topic.

My concern is deduping a retained packet only on CONNECT.

**Comment by mcollina** [comment]
by deduping you mean the overlapping subscription case?

**Comment by behrad** [comment]
No, I mean deduping on pubacked (packet, clientId) pair, so that a packet is only delivered once to a clientId

**Comment by mcollina** [comment]
why are you getting duplicates?

**Comment by GavinDmello** [comment]
@behrad. Can you check if your client internally resubscribes after a reconnection ? If so, the behaviour is correct.

**Comment by behrad** [comment]
> Can you check if your client internally resubscribes after a reconnection ? If so, the behaviour is correct.

Yes it does, and as you said this is the correct behavior for that case ( i.e. CONNECT/SUBSCRIBE)
However I'm concerned with a CONNECT only case.

**Comment by behrad** [comment]
> why are you getting duplicates?

Client is receiving the retained packet on every CONNECT. I define this as duplicate

**Comment by mcollina** [comment]
So, I think we should just have an option to disable sending retained messages for clean = false clients.

**Comment by GavinDmello** [comment]
You could also try modifying your client to take an option with switches off the resubscribe and handle all the subs yourself. This way you won't get it on every connect.

**Comment by behrad** [comment]
But we don't expect `clean=false` clients not to receive retained messages, but also receive them only once @mcollina

**Comment by behrad** [comment]
> You could also try modifying your client to take an option with switches off the resubscribe and handle all the subs yourself. This way you won't get it on every connect.

I didn't get your point @GavinDmello . Even if I switch off resubscribes, my client will get the packet on every connect.

**Comment by GavinDmello** [comment]
@behrad Your client should not get the retained message if it doesn't actively subscribe/resubscribe. 
If you're getting a retained message by just connecting , something is wrong.

**Comment by behrad** [comment]
> Your client should not get the retained message if it doesn't actively subscribe/resubscribe.
> If you're getting a retained message by just connecting , something is wrong.

Aedes wise, this is the path: connect -> restoreSubs -> handleSubscribe -> completeSubscribe! and `createRetainedStream` is called there after CONNECT @GavinDmello 

MQTT spec wise, I hadn't have a chance to ask this in their list to find out whats the correct behavior?

**Comment by behrad** [comment]
So currently in aedes, client will receive retained message on every CONNECT.
if the client does a SUBSCRIBE after that, it will again receive retained message.

**Comment by mcollina** [comment]
I am not sure that disabling that flag would help. The way deduping is
implemented relies on the fact that two messages are published in a short
period of time. A retained message could be retrieved after a long period
of time.

You might get away with it by setting the brokerId to a special id, and
start numbering the packets from 0.

I'm ok if you want to send a PR to have deduping. However, I'd recommend
you to not do a SUBSCRIBE if your client has resumed the session.
Il giorno gio 13 lug 2017 alle 21:09 Behrad <notifications@github.com> ha
scritto:

**Comment by behrad** [comment]
> However, I'd recommend
> you to not do a SUBSCRIBE if your client has resumed the session

I won't do a SUB for clean=false clients.


> You might get away with it by setting the brokerId to a special id, and
> start numbering the packets from 0.

My issue is then is delivering retained messages on every CONNECT of a client. and I can't see how this would help yet @mcollina

**Comment by mcollina** [comment]
I'm really not understanding your use case @behrad, would you mind setting up an example on how things currently behave, and telling which messages you would not like to be delivered?

Or, fire a PR and we will discuss there.

**Comment by behrad** [comment]
suspended until Ive free time to think more since this is a non-standard behavior
